### PR TITLE
DevTest: Move generated textures to world dir

### PR DESCRIPTION
**Pull request by Wuzzy2**
I always had this annoying issue with DevTest on my computer by spitting out errors that the generated textures `testnodes_generated_*.png` could not be read. This is because DevTest is installed system-wide and DevTest tries to put the textures into the mod directory, but it's write-protected without root rights. (I'm on an Arch-Linux-like system btw).

So I've decided to move the generated textures into the world path instead. This works fine. The world path is a "safer" option anyway.

### CODE REVIEW

File: games/devtest/mods/testnodes/textures.lua [hunk 1]
```diff
@@ -149,15 +149,20 @@ fractal = nil
...
-local textures_path = minetest.get_modpath( minetest.get_current_modname() ) .. "/textures/"
-minetest.safe_file_write(
-	textures_path .. "testnodes_generated_mb.png",
-	encode_and_check(512, 512, "rgb", data_mb)
-)
-minetest.safe_file_write(
-	textures_path .. "testnodes_generated_ck.png",
-	encode_and_check(512, 512, "gray", data_ck)
-)
+local textures_path = minetest.get_worldpath() .. "/generated_textures/"
```
Comment (sfan5) [inline games/devtest/mods/testnodes/textures.lua:152]: but the engine won't read media from this path, so this doesn't work

**Comment by Desour** [comment]
Can't you just call `core.dynamic_add_media` on load time to add the paths?

**Comment by Wuzzy2** [comment]
Hmmm, I just experimented around with that function (see latest commit) but Minetest doesn't like it:

```
2024-01-17 15:11:05: ERROR[Server]: Server::dynamicAddMedia(): file "testnodes_generated_mb.png" already exists in media cache
2024-01-17 15:11:05: ERROR[Server]: Server::dynamicAddMedia(): file "testnodes_generated_ck.png" already exists in media cache
```

Also happens with new worlds. The textures display correctly, both in singleplayer and multiplayer but these errors are concerning.

I don't know how to call the `dynamic_add_media` function correctly, it seems. Why are these errors triggered? They say they already exist in media cache but how can this happen with a brand new world? I also tried to erase my local media cache, still the same.

Somehow these textures must have gotten into the media cache before the dynamic add media? But how? The documentation is sadly not of great help here either. :-(

If I understand correctly, textures generated and written into the world directory are *not* automatically sent to players, right?

**Comment by Desour** [comment]
> file "testnodes_generated_mb.png" already exists in media cache

Did you remove the generated files from the mod's textures folder?

Because the textures are used for nodes, sending after the player joined does not work.
My idea was to not set `to_player`, and set `ephemeral=false`, to send it like other media at join. I wanted to call it at load time, but was reminded by minetest that this is for some reason not allowed.
`on_prejoinplayer` works though (`core.after(0,fun)` also works, but idk if it's guaranteed that no player joins before the first globalstep):
https://github.com/Desour/minetest/tree/Wuzzy2_devtest_fix_gentexes_prejoin

I haven't figured out how to create a commit that deletes previously gitignored files.

**Comment by sfan5** [comment]
Probable cause: You didn't clean the old `testnodes_generated_mb.png` from inside games/devtest/ so it is being put into the media cache.

What you're doing won't work anyway however (you can again see that [in CI](https://github.com/minetest/minetest/actions/runs/7556954070/job/20575121345?pr=14253#step:5:204)).
The client sets up node textures before the join completes, if you send textures in `on_joinplayer` it's too late.

What you can do is this:
```lua
core.after(0, function()
     core.dynamic_add_media({filepath = insert_path_here}, function() end)
end)
```

**Comment by Wuzzy2** [comment]
Ah yes, this was it. I tried `minetest.register_on_mods_loaded` before but this was wrong. Also, it turns out I indeed still had the generated files in the mod folder. Fun fact: They were hidden from `git status` thanks to `.gitignore`. Sneaky sneaky.

This thing now seems to work but let's see what CI will say. \*sweats nervously\*

**Comment by Desour** [comment]
We still need to somehow with a commit remove the texture files in the mod's texture folder. Otherwise, when you git pull master, they'll be untracked by git, and also devtest will fail. This would be annoying and werid.

**Comment by Wuzzy2** [comment]
I'm not sure but I don't think it is possible to "remove" a file in Git that was never tracked before. At least not trivially.

Also, this will be a one-time annoyance. The easy fix is to just do `git status` (which people should do all the time anyway), discover the offending files and manually remove them.

Maybe a post in the News forum will help, unless Git actually offers a way to automate this.

**Comment by Desour** [comment]
I think it might work if we first make a commit that adds the files. But that would be somewhat stupid, because we didn't want these files in the git history in the first place, for some reason.
Another solution could maybe be use different texture names, and keep them gitignored. But this is also weird.
Idk.

**Comment by Wuzzy2** [comment]
Umm ... Is this PR officially rejected or has it been merged without us noticing?

If this PR was rejected, this means the bug I mentioned still stands (there is no formal issue yet).

**Comment by sfan5** [comment]
Rejected. https://github.com/minetest/minetest/pull/14298 fixes this.

**Comment by Wuzzy2** [comment]
Thank you.

**Pull request closed without merging**